src/Microsoft.Framework.Caching.SqlServer/SqlServerCache.cs

```diff
@@ -0,0 +1,478 @@
+
+            StartScanForExpiredCacheItems();
+        }
+
+        public async Task SetAsync(
+            [NotNull] string key,
+            [NotNull] byte[] value,
+            [NotNull] DistributedCacheEntryOptions options)
+        {
+            var utcNow = _systemClock.UtcNow;
+            var expirationInfo = GetExpirationInfo(utcNow, options);
+
+            using (var connection = new SqlConnection(_options.ConnectionString))
+            {
+                // Try updating the cache item and if it does not succeed then its probably a new item
+                var command = new SqlCommand(_sqlQueries.UpdateCacheItem, connection);
+                command.Parameters.AddWithValue(CacheItemId, key);
+                command.Parameters.AddWithValue(CacheItemValue, value);
+                command.Parameters.AddWithValue(ExpiresAtTimeUTC, expirationInfo.ExpiresAtTimeUTC);
+
+                await connection.OpenAsync();
+
+                var recordsAffected = await command.ExecuteNonQueryAsync();
+
+                // only add if update didn't affect any records
+                if (recordsAffected <= 0)
+                {
+                    command = new SqlCommand(_sqlQueries.AddCacheItem, connection);
+                    command.Parameters.AddWithValue(CacheItemId, key);
+                    command.Parameters.AddWithValue(CacheItemValue, value);
+                    command.Parameters.AddWithValue(ExpiresAtTimeUTC, expirationInfo.ExpiresAtTimeUTC);
+
+                    if (expirationInfo.SlidingExpiration.HasValue)
+                    {
+                        command.Parameters.AddWithValue(
+                            SlidingExpirationInTicks,
+                            expirationInfo.SlidingExpiration.Value.Ticks);
+                    }
+                    else
+                    {
+                        command.Parameters.AddWithValue(SlidingExpirationInTicks, DBNull.Value);
+                    }
+
+                    if (expirationInfo.AbsoluteExpiration.HasValue)
+                    {
+                        command.Parameters.AddWithValue(
+                            AbsoluteExpiration,
+                            expirationInfo.AbsoluteExpiration.Value);
+                    }
+                    else
+                    {
+                        command.Parameters.AddWithValue(AbsoluteExpiration, DBNull.Value);
+                    }
+
+                    recordsAffected = await command.ExecuteNonQueryAsync();
+                }
+            }
+
+            StartScanForExpiredCacheItems();
+        }
+
+        // Called by multiple actions to see how long it's been since we last checked for expired items.
+        // If sufficient time has elapsed then a scan is initiated on a background task.
+        private void StartScanForExpiredCacheItems()
+        {
+            var utcNow = _systemClock.UtcNow;
+            if ((utcNow - _lastExpirationScan) > _options.ExpirationScanFrequency)
```

Comment:
Can the interval be low enough (i.e. the frequency high enough) that you will have more than one worker thread doing this at the same time? It is possibly just a corner case but if there was a simple and nice pattern to guarantee that there is only one doing it at a time I would probably prefer that. Otherwise we should consider validating that this ExpirationScanFrequency has a reasonable value.